app/models/zbmath.rb

```diff
@@ -0,0 +1,192 @@
+  end
+
+  def source_id
+    "zbmath"
+  end
+
+  def get_records(options = {})
+    client = OAI::Client.new "https://oai.portal.mardi4nfdi.de/oai/OAIHandler"
+    begin
+      response = client.list_records metadata_prefix: "datacite_articles", from: options[:from],
+                                     until: options[:until]
+
+      # Get the metadata - read_datacite expects a string of the XML tree with the <resource> element as the root,
+      # and OAI wraps the data in a <metadata> element so strip this with string slicing (a little ugly, but a lot
+      # simpler and more efficient than parsing the XML, restructuring the tree and then serializing back to string)
+      response.each do |record|
+        ZbmathImportJob.perform_later(record.metadata.to_s[12..-14])
+      end
+      response.count
+    rescue OAI::NoMatchException
+      Rails.logger.info "No ZBMath records updated between #{options[:from]} and #{options[:until]}."
+      nil
+    end
+  end
+
+  def self.parse_zbmath_record(record)
+    meta = read_datacite(string: record)
+
+    # Check that the record is for a DataCite DOI
+    doi = meta.fetch("doi", nil)
+    return nil if doi.blank? # && cached_doi_ra(doi) == "DataCite"
+
+    pid = normalize_doi(doi)
+    subj = cached_datacite_response(pid)
+    # parse out valid related identifiers
+    related_doi_identifiers = Array.wrap(meta.fetch("related_identifiers", nil)).select do |r|
+      %w(DOI URL).include?(r["relatedIdentifierType"]) && r["relatedIdentifier"] != "https://zbmath.org"
+    end
+
+    # loop through related identifiers and build event objects
+    items = Array.wrap(related_doi_identifiers).reduce([]) do |x, item|
+      related_identifier = item.fetch("relatedIdentifier", nil).to_s.strip.downcase
+      related_identifier_type = item.fetch("relatedIdentifierType", nil).to_s.strip
+
+      if related_identifier_type == "DOI"
+        obj_id = normalize_doi(related_identifier)
+
+        # Get RA and populate obj
+        ra = cached_doi_ra(related_identifier)
+        obj = if ra == "DataCite"
+                cached_datacite_response(obj_id)
+              elsif ra == "Crossref"
+                cached_crossref_response(obj_id)
+              else
+                {}
+              end
+      else
+        # It's a URL so no PID object on the other end
+        obj_id = normalize_url(related_identifier)
+        obj = {}
+      end
+
+      x << { "message_action" => "create",
+             "id" => SecureRandom.uuid,
+             "subj_id" => pid,
+             "obj_id" => obj_id,
+             "relation_type_id" => item["relationType"].to_s.underscore,
+             "source_id" => "zbmath_related",
+             "occurred_at" => Time.zone.now.utc,
+             "timestamp" => Time.zone.now.iso8601,
+             "license" => LICENSE,
+             "subj" => subj,
+             "obj" => obj,
+             "source_token" => "4c891c31-519e-4d98-ac8d-876ad0f28635" }
```

Comment:
Were you going to move this to an ENV variable for this release?

Reply:
Yes - I probably will do this, although I still haven't worked out what we use it for, or why it can't just be hardcoded :sweat_smile:

Reply:
Done!